### PR TITLE
CI: Fix most mypy errors in the Dask backend

### PR DESCRIPTION
**Pull request by sofide**
Partially solves issue #2852 

There are some remaining errors, but I think they might need a deeper code refactor.

### CODE REVIEW

File: ibis/backends/dask/execution/selection.py [hunk 1]
```diff
@@ -37,7 +37,7 @@ def compute_projection_scalar_expr(
     expr,
     parent,
     data,
-    scope: Scope = None,
+    scope: Scope,
```
Comment (datapythonista) [inline ibis/backends/dask/execution/selection.py:40]: Any reason to remove the default value, instead of setting the type to `Optional[Scope]`, which seems like a most conservative approach?

Comment (sofide) [reply ibis/backends/dask/execution/selection.py:40]: @datapythonista  The problem wasn't the missing type `Optional`: when a default is `None` the type `Optional` is implicitly added, and mypy doesn't have any problem with it (although explicit is always better than implicit :P )

The problem is the line 52:
```python
    scope = scope.merge_scopes(
```
If `scope` is `None` it will raise an exception.

Comment (datapythonista) [reply ibis/backends/dask/execution/selection.py:40]: Ah, great, thanka for the info. @icexelloss can you have a quick look at this PR please?

File: ibis/backends/dask/execution/util.py [hunk 2]
```diff
@@ -158,9 +160,7 @@ def _pandas_dtype_from_dd_scalar(x: dd.core.Scalar):
 
 
 def _coerce_to_dataframe(
-    data: Any,
-    column_names: List[str],
-    types: Optional[List[dt.DataType]] = None,
```
Comment (sofide) [inline ibis/backends/dask/execution/util.py:163]: @datapythonista The problem here is that if `types` is `None` and `data` is a `Series`, this part of the code will raise an error:
```python
    elif isinstance(data, dd.Series):
        dtypes = map(ibis_dtype_to_pandas, types)
```
Would you prefer to manage that case and raise a specific error?

File: ibis/backends/dask/client.py [hunk 3]
```diff
@@ -244,19 +244,18 @@ def create_table(
         schema: sch.Schema = None,
     ):
         """Create a table."""
-        if obj is None and schema is None:
-            raise com.IbisError('Must pass expr or schema')
-
         if obj is not None:
             df = obj
-        else:
+        elif schema is not None:
```
Comment (jreback) [inline ibis/backends/dask/client.py:249]: why this change? this is a logic change

**Review by jreback — CHANGES_REQUESTED** [review]
cc @gerrymanoim if you'd have a look

File: ibis/backends/dask/client.py (hunk 3 above)
Comment (datapythonista) [reply ibis/backends/dask/client.py:249]: Not sure if there was any mypy problem in the original code, but the logic is the same (not immediately obvious, took me a bit to see it), and I think the new implementation is clearer.

Comment (sofide) [reply ibis/backends/dask/client.py:249]: Without this change, mypy raise this error:
```
ibis/backends/dask/client.py:254: error: Item "None" of "Optional[Schema]" has no attribute "apply_to"
```
The logic is the same, but in the original code mypy cannot figure out that schema will never be `None` in that line.

Comment (gerrymanoim) [reply ibis/backends/dask/client.py:249]: Shouldn't this look the same in pandas then? https://github.com/ibis-project/ibis/blob/master/ibis/backends/pandas/client.py#L409

I'm fine with the change, but I'd expect that the `PandasClient` and `DaskClient` look extremely similar code-wise.

File: ibis/backends/dask/execution/util.py (hunk 2 above)
Comment (gerrymanoim) [reply ibis/backends/dask/execution/util.py:163]: I'm fine removing the optional. This is only used internally in `coerce_to_output` which will always pass types.

File: ibis/backends/dask/client.py (hunk 3 above)
Comment (datapythonista) [reply ibis/backends/dask/client.py:249]: I created issues to address one backend at a time. But we'll surely have to make the same change in the pandas backend when we fix the mypy errors on it, since the mypy problem will be the same.

**Review by datapythonista — APPROVED** [review]
Thanks @sofide, for the contribution, and for all the additional information. All good then. @jreback if you want to check the comments...

File: ibis/backends/dask/client.py (hunk 3 above)
Comment (sofide) [reply ibis/backends/dask/client.py:249]: @arielrossanigo have taken this issue in the pandas backend, maybe we can coordinate to make the same change in both backends

Comment (arielrossanigo) [reply ibis/backends/dask/client.py:249]: Hi people, I've checked and currently in the `PandasClient` there is no mypy problem because of the lack of annotations. 

I think @sofide can reflect the change in the `PandasClient` and then I rebase my branch when this PR is merged, What do you think?

Comment (datapythonista) [reply ibis/backends/dask/client.py:249]: Don't worry about this, this PR should be ready. Will wait a bit more to see if @jreback can have a last look, but all is fine.

And regarding pandas, not so concerned that this function has the same code. I don't think we should have copied pasted anyway, so if we want to do things properly what we should do is use the pandas function for dask, not make sure we make the same changes in both functions. But let's focus on the mypy problems.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @sofide
